NodeDependencyLookup/Editor/Caches/CacheSerializerUtils.cs

```diff
@@ -1,143 +1,214 @@
-                EncodeString(dependency.DependencyType, ref bytes, ref offset);
-                EncodeString(dependency.NodeType, ref bytes, ref offset);
-
-                EncodePathSegments(dependency.PathSegments, ref bytes, ref offset);
-
-                bytes = EnsureSize(bytes, offset);
-            }
-        }
-        
-        public static List<Dependency> DecodeDependencies(ref byte[] bytes, ref int offset)
-        {
-            int numDependencies = DecodeShort(ref bytes, ref offset);
-            List<Dependency> dependencies = new List<Dependency>();
-
-            for (var k = 0; k < numDependencies; k++)
-            {
-                string id = DecodeString(ref bytes, ref offset);
-                string connectionType = DecodeString(ref bytes, ref offset);
-                string nodeType = DecodeString(ref bytes, ref offset);
-                PathSegment[] pathSegments = DecodePathSegments(ref bytes, ref offset);
-							
-                dependencies.Add(new Dependency(id, connectionType, nodeType, pathSegments));
-            }
-
-            return dependencies;
-        }
-    }
+	/// <summary>
+	/// Contains all necessary functions to encode and decode the cache to binary and back
+	/// Reason for doing own conversion instead of using json is speed.
+	/// </summary>
+	public static class CacheSerializerUtils
+	{
+		public const int ARRAY_SIZE_OFFSET = 0xFFFF; // 64 kb
+
+		public static void EncodeShort(short value, ref byte[] bytes, ref int offset)
+		{
+			bytes[offset++] = (byte) value;
+			bytes[offset++] = (byte) (value >> 8);
+		}
+
+		public static short DecodeShort(ref byte[] bytes, ref int offset)
+		{
+			return (short) (bytes[offset++] + (bytes[offset++] << 8));
+		}
+
+		public static void EncodeLong(long value, ref byte[] bytes, ref int offset)
+		{
+			for (var k = 0; k < 8; ++k)
+			{
+				bytes[offset++] = (byte) (value >> (8 * k));
+			}
+		}
+
+		public static long DecodeLong(ref byte[] bytes, ref int offset)
+		{
+			long result = 0;
+
+			for (var k = 0; k < 8; ++k)
+			{
+				result += (long) bytes[offset++] << (8 * k);
+			}
+
+			return result;
+		}
+
+		public static byte[] EnsureSize(byte[] array, int offset)
+		{
+			if (offset + ARRAY_SIZE_OFFSET / 2 > array.Length)
+			{
+				var newArray = new byte[array.Length * 2];
+				Array.Copy(array, newArray, offset);
+				return newArray;
+			}
+
+			return array;
+		}
+
+		public static void EncodeString(string value, ref byte[] bytes, ref int offset)
+		{
+			var charArray = value.ToCharArray();
+			EncodeShort((short) charArray.Length, ref bytes, ref offset);
+
+			for (var c = 0; c < charArray.Length; c++)
+			{
+				bytes[offset++] = (byte) charArray[c];
+			}
+		}
+
+		public static string DecodeString(ref byte[] bytes, ref int offset)
+		{
+			int length = DecodeShort(ref bytes, ref offset);
+
+#if UNITY_2021_1_OR_NEWER
+            Span<char> charArray = stackalloc char[length];

```

Comment:
What's the benefit of this?

Reply:
Less garbage created since I dont need to allocate char[] on the heap. Instead it is done on the stack.